"step2

nekisse-lee opened the pull request:
피드백 감사합니다.

nekisse-lee committed:
"step2 QnaService update, delete 기능구현

"

nekisse-lee committed:
"step2 question관련 기능 구현 및 단위테스트, ATDD테스트 작성
       answer 기능은 구현해야함.

- HtmlFormDataBuilder.class
   post, put 메서드추가,

- QuestionAcceptanceTest.class
    Question에 대한 테스트만 구현

- import.sql
   테스트데이터 추가

- updateForm.html
    hidden put으로  추가

"

jihan805 commented on src/test/java/support/test/HtmlFormDataBuilder.java:
```diff
@@ -23,6 +23,16 @@ public HtmlFormDataBuilder addParameter(String key, Object value) {
         return this;
     }
 
+    public HtmlFormDataBuilder delete() {
```
delete 와 post 메서드 구현 👍

jihan805 commented on src/main/java/nextstep/service/QnaService.java:
```diff
@@ -36,14 +38,20 @@ public Question create(User loginUser, Question question) {
+        Question question = findById(id).orElseThrow(UnAuthenticationException::new);
+        question.update(loginUser, updatedQuestion);
+        return questionRepository.save(question);
     }
```
메소드에 @ Transactional 애노테이션이 있습니다! questionRepository.save(..)와 같이 명시적으로 구현하지 않고 단순히 question.update(questionDto) 만으로도 수정이 가능한지 테스트해 보면 어떨까요!?

jihan805 commented on src/main/java/nextstep/web/LoginController.java:
```diff
@@ -0,0 +1,42 @@
+
+@Controller
+
+public class LoginController {
```
테스트 코드를 포함하여 전체적으로 불필요하게 줄을 띄우지 않아도 좋을 것 같아요!

jihan805 commented on src/main/java/nextstep/domain/Question.java:
```diff
@@ -60,6 +62,14 @@ public User getWriter() {
     }
 
+    public void update(User loginUser, Question updatedQuestion) throws UnAuthenticationException {
+        if (!isOwner(loginUser)) {
```
Question에서 자신이 쓴 글인지 확인하도록 구현 👍

jihan805 commented on src/test/java/nextstep/domain/QuestionTest.java:
```diff
@@ -0,0 +1,34 @@
+import support.test.BaseTest;
+
+public class QuestionTest extends BaseTest {
+    private static final User LOGIN_USER = new User(1L, "nekisse", "password1", "nname", "nekisse@c.com");
```
QuestionTest와 QuestionAcceptanceTest 구현 👍 
앞서 UserService에 대한  Test가 구현되어 있을텐데요, QnaServiceTest도 구현하는 연습을 해보면 어떨까요!?

jihan805 commented on src/main/java/nextstep/service/QnaService.java:
```diff
@@ -36,14 +38,20 @@ public Question create(User loginUser, Question question) {
+            throw new UnAuthenticationException("자신의 글만 삭제 가능합니다.");
+        }
+         questionRepository.delete(question);
     }
```
해당 부분은 4단계 요구사항으로 나오는 부분이기도 합니다!
현재는 questionRepository를 통해 물리적으로 삭제하고 있어요.
Question 클래스의 private boolean deleted 라는 필드가 있는데 해당 필드를 이용해 보는 것은 어떨까요!?
내가 쓴 글인지를 체크하는 로직도 재활용할 수 있을 것 같아요.

jihan805 commented on src/main/java/nextstep/service/QnaService.java:
```diff
@@ -56,6 +64,7 @@ public void deleteQuestion(User loginUser, long questionId) throws CannotDeleteE
 
     public Answer addAnswer(User loginUser, long questionId, String contents) {
         // TODO 답변 추가 기능 구현
```
2단계에서는 질문에 대한 CRUD 구현이므로
답변에 대한 구현은 3단계 (질문과 답변 기능에 대한 RESTful API를 제공한다.)에서 구현하셔도 될것같아요!

jihan805 reviewed (changes requested):
안녕하세요! qna리뷰를 담당하게 된 백지한입니다!
2단계 구현 잘하셨어요 👍 
몇가지 피드백 남겼으니 확인 부탁드릴게요.
감사합니다 :)

nekisse-lee committed:
"step2 피드백 반영사항 수정
- Question.class
    delete()메서드 추가

- QnaService.class
    deleteQuestion() 메서드 수정
    update() 메서드 수정.

- LoginController.class
    공백제거
- QnaServiceTest.class
    테스트작성

jihan805 reviewed (approved):
안녕하세요~! 피드백 반영 감사합니다!
머지 진행할게요 👍 
다음단계 진행 부탁드리겠습니다. 
감사합니다 :)
